File: packages/detector-core/aiprojectdetector/detectors/llm_fingerprint.py
"""LLM fingerprint heuristics: tell-tale comments, placeholders, over-docs.

These are the signals most specific to *generated* code as opposed to merely
clean code: tutorial-style narration in comments, placeholder stubs, generic
TODOs, emoji in source comments, and comment-to-code ratios far above what
working engineers typically write.
"""

from __future__ import annotations

import re

from ..constants import LLM_COMMENT_PHRASES, PLACEHOLDER_MARKERS
from ..models import Signal
from ..parsing.languages import strip_comments
from ..utils.text import clamp
from .base import AnalysisUnit, Detector

_EMOJI_RE = re.compile(
    "[\U0001F300-\U0001FAFF\U00002600-\U000027BF\U0001F1E6-\U0001F1FF✀-➿]"
)
_TODO_RE = re.compile(r"\b(todo|fixme|xxx|hack)\b", re.IGNORECASE)


class LLMFingerprintDetector(Detector):
    name = "llm_fingerprint"
    description = "Tell-tale LLM phrasings, placeholders, generic TODOs, over-documentation."

    def unit_signals(self, unit: AnalysisUnit) -> list[Signal]:
        if unit.is_config or unit.loc < 5:
            return []
        code, comments = strip_comments(unit.source, unit.language)
        comment_text = "\n".join(comments).lower()
        comment_lines = [c for c in comments if c.strip()]
        code_lines = [ln for ln in code.split("\n") if ln.strip()]
        signals: list[Signal] = []

        # 1) Generic / tutorial-style comments.
        if comment_lines:
            hits = [p for p in LLM_COMMENT_PHRASES if p in comment_text]
            density = len(hits) / max(1, len(comment_lines))
            score = clamp(0.5 + 0.45 * clamp(density / 0.4) + 0.1 * clamp(len(hits) / 6))
            ev = []
            if hits:
                sample = self._first_comment_with(comments, hits)
                ev.append(
                    self.evidence(
                        "generic_comments",
                        f"Comments use {len(hits)} phrasing(s) common in LLM output "
                        f"(e.g. {', '.join(repr(h) for h in hits[:4])}).",
                        severity=score,
                        path=unit.path,
                        start_line=unit.start_line,
                        snippet=sample,
                    )
                )
            signals.append(
                self.signal(
                    "generic_comments",
                    score,
                    confidence=clamp(len(comment_lines) / 15),
                    weight=1.3,
                    reason=f"{len(hits)} LLM-typical comment phrase(s) over "
                    f"{len(comment_lines)} comment line(s).",
                    evidence=ev,
                )
            )

        # 2) Placeholder / stub implementations.
        lower_all = unit.source.lower()
        ph_hits = [m for m in PLACEHOLDER_MARKERS if m in lower_all]
        if ph_hits:
            score = clamp(0.55 + 0.4 * clamp(len(ph_hits) / 4))
            signals.append(
                self.signal(
                    "placeholder_implementations",
                    score,
                    confidence=0.55,
                    weight=1.1,
                    reason=f"Found placeholder/stub markers: {', '.join(repr(h) for h in ph_hits[:4])}.",
                    evidence=[
                        self.evidence(
                            "placeholder_implementations",
                            f"Placeholder/stub text present ({', '.join(ph_hits[:3])}).",
                            severity=score,
                            path=unit.path,
                            start_line=unit.start_line,
                        )
                    ],
                )
            )

        # 3) Over-documentation: comment-to-code ratio far above typical.
        if len(code_lines) >= 10:
            ratio = len(comment_lines) / max(1, len(code_lines))
            score = clamp(0.5 + 0.4 * clamp((ratio - 0.35) / 0.6))
            if ratio > 0.45:
                signals.append(
                    self.signal(
                        "excessive_documentation",
                        score,
                        confidence=clamp(len(code_lines) / 60),
                        weight=0.8,
                        reason=f"Comment-to-code ratio {ratio:.0%} (high).",
                        evidence=[
                            self.evidence(
                                "excessive_documentation",
                                f"Unusually high comment-to-code ratio ({ratio:.0%}).",
                                severity=score,
                                path=unit.path,
                            )
                        ],
                    )
                )

        # 4) Emoji in source comments.
        emoji = _EMOJI_RE.findall(comment_text)
        if emoji:
            score = clamp(0.55 + 0.3 * clamp(len(emoji) / 4))
            signals.append(
                self.signal(
                    "emoji_in_comments",
                    score,
                    confidence=0.5,
                    weight=0.6,
                    reason=f"{len(emoji)} emoji found in code comments.",
                )
            )

        # 5) Generic TODOs without specifics.
        todos = _TODO_RE.findall(comment_text)
        if todos:
            generic = sum(1 for c in comments if _TODO_RE.search(c) and len(c.strip()) < 24)
            if generic:
                score = clamp(0.5 + 0.25 * clamp(generic / 4))
                signals.append(
                    self.signal(
                        "generic_todos",
                        score,
                        confidence=0.45,
                        weight=0.5,
                        reason=f"{generic} short/generic TODO-style comment(s).",
                    )
                )

        return signals

    @staticmethod
    def _first_comment_with(comments: list[str], phrases: list[str]) -> str | None:
        for c in comments:
            low = c.lower()
            if any(p in low for p in phrases):
                return c.strip()[:200]
        return None

File: packages/detector-core-ts/src/detectors/llmFingerprint.ts
/**
 * LLM fingerprint heuristics: tell-tale comments, placeholders, over-docs.
 *
 * These are the signals most specific to *generated* code as opposed to merely
 * clean code: tutorial-style narration, placeholder stubs, generic TODOs,
 * emoji in comments, and unusually high comment-to-code ratios.
 */

import { LLM_COMMENT_PHRASES, PLACEHOLDER_MARKERS } from "../constants.js";
import type { Signal } from "../models.js";
import { stripComments } from "../parsing/languages.js";
import { clamp } from "../utils/text.js";
import { percent0 } from "../utils/format.js";
import { Detector, unitLoc, type AnalysisUnit } from "./base.js";

// Mirror of the Python emoji character ranges.
export const EMOJI_RE =
  /[\u{1F300}-\u{1FAFF}\u{2600}-\u{27BF}\u{1F1E6}-\u{1F1FF}\u{2740}-\u{27BF}]/gu;
const TODO_RE = /\b(todo|fixme|xxx|hack)\b/i;
const TODO_RE_G = /\b(todo|fixme|xxx|hack)\b/gi;

function pyRepr(s: string): string {
  // Mirror Python repr() for the common single-quoted-string case used here.
  return `'${s.replace(/\\/g, "\\\\").replace(/'/g, "\\'")}'`;
}

export class LLMFingerprintDetector extends Detector {
  override name = "llm_fingerprint";
  override description =
    "Tell-tale LLM phrasings, placeholders, generic TODOs, over-documentation.";

  override unitSignals(unit: AnalysisUnit): Signal[] {
    if (unit.is_config || unitLoc(unit) < 5) return [];
    const [code, comments] = stripComments(unit.source, unit.language);
    const commentText = comments.join("\n").toLowerCase();
    const commentLines = comments.filter((c) => c.trim());
    const codeLines = code.split("\n").filter((ln) => ln.trim());
    const signals: Signal[] = [];

    // 1) Generic / tutorial-style comments.
    if (commentLines.length) {
      const hits = LLM_COMMENT_PHRASES.filter((p) => commentText.includes(p));
      const density = hits.length / Math.max(1, commentLines.length);
      const score = clamp(0.5 + 0.45 * clamp(density / 0.4) + 0.1 * clamp(hits.length / 6));
      const ev = [];
      if (hits.length) {
        const sample = this.firstCommentWith(comments, hits);
        ev.push(
          this.evidence(
            "generic_comments",
            `Comments use ${hits.length} phrasing(s) common in LLM output ` +
              `(e.g. ${hits.slice(0, 4).map(pyRepr).join(", ")}).`,
            { severity: score, path: unit.path, start_line: unit.start_line, snippet: sample },
          ),
        );
      }
      signals.push(
        this.signal("generic_comments", score, {
          confidence: clamp(commentLines.length / 15),
          weight: 1.3,
          reason:
            `${hits.length} LLM-typical comment phrase(s) over ` +
            `${commentLines.length} comment line(s).`,
          evidence: ev,
        }),
      );
    }

    // 2) Placeholder / stub implementations.
    const lowerAll = unit.source.toLowerCase();
    const phHits = PLACEHOLDER_MARKERS.filter((m) => lowerAll.includes(m));
    if (phHits.length) {
      const score = clamp(0.55 + 0.4 * clamp(phHits.length / 4));
      signals.push(
        this.signal("placeholder_implementations", score, {
          confidence: 0.55,
          weight: 1.1,
          reason: `Found placeholder/stub markers: ${phHits.slice(0, 4).map(pyRepr).join(", ")}.`,
          evidence: [
            this.evidence(
              "placeholder_implementations",
              `Placeholder/stub text present (${phHits.slice(0, 3).join(", ")}).`,
              { severity: score, path: unit.path, start_line: unit.start_line },
            ),
          ],
        }),
      );
    }

    // 3) Over-documentation: comment-to-code ratio far above typical.
    if (codeLines.length >= 10) {
      const ratio = commentLines.length / Math.max(1, codeLines.length);
      const score = clamp(0.5 + 0.4 * clamp((ratio - 0.35) / 0.6));
      if (ratio > 0.45) {
        signals.push(
          this.signal("excessive_documentation", score, {
            confidence: clamp(codeLines.length / 60),
            weight: 0.8,
            reason: `Comment-to-code ratio ${percent0(ratio)} (high).`,
            evidence: [
              this.evidence(
                "excessive_documentation",
                `Unusually high comment-to-code ratio (${percent0(ratio)}).`,
                { severity: score, path: unit.path },
              ),
            ],
          }),
        );
      }
    }

    // 4) Emoji in source comments.
    const emoji = commentText.match(EMOJI_RE);
    if (emoji && emoji.length) {
      const score = clamp(0.55 + 0.3 * clamp(emoji.length / 4));
      signals.push(
        this.signal("emoji_in_comments", score, {
          confidence: 0.5,
          weight: 0.6,
          reason: `${emoji.length} emoji found in code comments.`,
        }),
      );
    }

    // 5) Generic TODOs without specifics.
    const todos = commentText.match(TODO_RE_G);
    if (todos && todos.length) {
      const generic = comments.filter((c) => TODO_RE.test(c) && c.trim().length < 24).length;
      if (generic) {
        const score = clamp(0.5 + 0.25 * clamp(generic / 4));
        signals.push(
          this.signal("generic_todos", score, {
            confidence: 0.45,
            weight: 0.5,
            reason: `${generic} short/generic TODO-style comment(s).`,
          }),
        );
      }
    }

    return signals;
  }

  private firstCommentWith(comments: string[], phrases: string[]): string | null {
    for (const c of comments) {
      const low = c.toLowerCase();
      if (phrases.some((p) => low.includes(p))) return c.trim().slice(0, 200);
    }
    return null;
  }
}
